### DISCUSSION TITLE
Complete to selection while preserving candidates

### DISCUSSION THREAD

**Discussion by countvajhula** [company-mode/company-mode#1327] [Ideas]
Due to recent candidates being placed at the top (via [prescient](https://github.com/radian-software/prescient.el), in my case), I sometimes find something like this:
```
su|
-> subprocess [recent]
-> sub1
-> super
-> submod
-> suffix
...
```
.. a very long list. But in this case, I often do want to leverage the recency information, e.g. completing to the top one, which would then narrow the search space:

```
subprocess|
-> subprocess
-> subprocess?
-> subprocess-pid
-> subprocess-wait
...
```
which is the relevant narrowed list that I'd like to continue doing completion from (and for instance, I might pick `subprocess-wait` here).

For this, I'd need something resembling `company-complete-selection` but which doesn't "finish". I wrote this function that works:

```
(defun my-company-complete-to-selection ()
  "Insert the selected candidate but retain the completion menu."
  (interactive)
  (when (and (company-manual-begin) company-selection)
    (let ((result (nth company-selection company-candidates)))
      (if (and (not (cdr company-candidates))
               (equal result (car company-candidates)))
          (company-complete-selection)
        (company--insert-candidate result)))))
```

And I've bound it as S-tab in the `use-package` declaration:

```
  :bind (:map company-active-map
              ("<tab>" . #'company-complete-common)
              ("S-<tab>" . #'my-company-complete-to-selection))
```

**Comment by dgutov** [comment]
I take it `company-complete-common` does not fit your requirements?

**Reply by countvajhula** [reply]
That's right -- I do use `company-complete-common`, but in the example I posted earlier, `su->` does not complete since there are many candidate completions that start with `su` -- `sub1`, `super`, etc. Another way to look at it is that the main way in which company mode allows us to narrow the search tree is by typing additional characters. With the above feature, it would also allow us to narrow the search tree by selection.

**Reply by yugaego** [reply]
> With the above feature, it would also allow us to narrow the search tree by selection.

I'm curious if you had a chance to try out `M-x company-complete-quick-access`. The default keybinding `M-1` *might* suit the use case you describe, completing with the first candidate. Though that would probably not be ideal by default, since `company-complete-quick-access` also calls `company-finish` that seems made the difference for you.

And, considering `"narrow the search tree"` more abstractly, I'd like to mention -- for the possible future readers -- `M-x company-search-candidates`, bound to `C-s` by default,  and `M-x company-filter-candidates`, bound to `C-M-s`.

**Reply by countvajhula** [reply]
I tried `company-complete-quick-access` -- good to know about this, but it does not fit my usecase since, as you said, the completion menu disappears. I think that makes this a way to select a completion leaf rather than a way to select a completion subtree.

I'm not familiar with `company-search`, and `C-s`, `C-r`, `C-o` and `C-M-s` don't seem to do anything. I noticed this in the `*Messages*` buffer `user-error: Empty search string [17 times]`. How does one set the search string?

**Reply by countvajhula** [reply]
Ah okay, I didn't realize I was just supposed to keep typing. This is a cool feature! It's not exactly what I'm looking for though wrt subtree selection.

**Reply by yugaego** [reply]
> I think that makes this a way to select a completion leaf rather than a way to select a completion subtree.

Yup.

> Ah okay, I didn't realize I was just supposed to keep typing. This is a cool feature!

I'm glad you liked it. I specifically find the filtering (`C-M-s` by default) helpful, since it filters out all the candidates that don't match additionally typed characters.
